Changing AKS load_balancer_profile settings after creation (#5846)

aristosvo opened the pull request:
Fixes #5846 

## Issue
When the `load_balancer_profile` settings for `service/kubernetes-cluster` are all computed, changing (for example) from managed IPs to provided outbound IP addresses is not possible with only TerraForm without recreating your cluster.

## Changes
With this change applied
- `effective_outbound_ips` -> `Computed: true` (no change)
- `outbound_ip_address_ids` -> `Computed: false`
- `outbound_ip_prefix_ids` -> `Computed: false`
- `managed_outbound_ip_count` -> `Computed: false`

As only one of the three options (`outbound_ip_address_ids`, `outbound_ip_prefix_ids`, `outbound_ip_prefix_ids` ) could be set and at least one of them should be set, `Computed: true` doesn't make sense to me for the last three and it prevented changes.

If there is a better option to make a change possible, let me know

katbyte commented on azurerm/internal/services/containers/resource_arm_kubernetes_cluster.go:
```diff
@@ -298,14 +298,12 @@ func resourceArmKubernetesCluster() *schema.Resource {
 									"outbound_ip_prefix_ids": {
 										Type:          schema.TypeSet,
 										Optional:      true,
-										Computed:      true,
```
You should be able to set this to an empty array:
```
outbound_ip_prefix_ids = []
```

katbyte commented on azurerm/internal/services/containers/resource_arm_kubernetes_cluster.go:
```diff
@@ -316,7 +314,6 @@ func resourceArmKubernetesCluster() *schema.Resource {
 									"outbound_ip_address_ids": {
 										Type:          schema.TypeSet,
 										Optional:      true,
-										Computed:      true,
```
same here

katbyte commented on azurerm/internal/services/containers/resource_arm_kubernetes_cluster.go:
```diff
@@ -298,14 +298,12 @@ func resourceArmKubernetesCluster() *schema.Resource {
 									"managed_outbound_ip_count": {
 										Type:          schema.TypeInt,
 										Optional:      true,
-										Computed:      true,
```
Another option might be to allow this to be set to 0 and then in that case send nothing/0?

aristosvo commented:
> could you describe a little more why the cluster needs to be recreated when `computed=true`?

@katbyte Thanks for your remarks! I've added a test to describe better what I want. `Compute: true` and `ConflictsWith` are both nice to keep around, but not when we don't have logic to prefer _desired_ state over _computed_ state.

Is there a nice way to:
- Check whether loadbalancer config fields are from the _desired_ state or _computed_
- Use _desired_ state over _computed_ when it is set?
If you'd have an example I could probably implement it myself. 

Thanks a lot anyway!

aristosvo commented:
@katbyte Can you give it another review? I have reworked it a few times and I've finaly a solution which works for me:) With the solution I implemented we can keep both `Computed: true` and the `ConflictsWith` rules in the schema.

I'm still learning a lot about Go/TF and enjoying it, don't hold back if you've any improvements!:)

aristosvo commented:
If I can find the time I'll try to solve the conflicts

tombuildsstuff commented:
@aristosvo awesome, thanks for rebasing this 👍

tombuildsstuff commented on azurerm/internal/services/containers/tests/resource_arm_kubernetes_cluster_network_test.go:
```diff
@@ -485,6 +486,60 @@ func testAccAzureRMKubernetesCluster_prefixedLoadBalancerProfile(t *testing.T) {
+					resource.TestCheckResourceAttr(data.ResourceName, "network_profile.0.load_balancer_profile.0.outbound_ip_prefix_ids.#", "1"),
+					resource.TestCheckResourceAttr(data.ResourceName, "network_profile.0.load_balancer_profile.0.effective_outbound_ips.#", "1"),
+				),
+			}
```
the compilation is failing here because of a missing `,`:


can we also add an import test between these two steps too, to check all the fields in the config are present in the state?

```suggestion
			},
			data.ImportStep(),
```

tombuildsstuff commented on azurerm/internal/services/containers/tests/resource_arm_kubernetes_cluster_network_test.go:
```diff
@@ -485,6 +486,60 @@ func testAccAzureRMKubernetesCluster_prefixedLoadBalancerProfile(t *testing.T) {
+					resource.TestCheckResourceAttr(data.ResourceName, "network_profile.0.load_balancer_profile.0.managed_outbound_ip_count", "1"),
+					resource.TestCheckResourceAttr(data.ResourceName, "network_profile.0.load_balancer_profile.0.effective_outbound_ips.#", "1"),
+				),
+			}
```
the compilation is failing here because of a missing `,`:


can we also add an import test between these two steps too, to check all the fields in the config are present in the state?

```suggestion
			},
			data.ImportStep(),
```

tombuildsstuff commented on azurerm/internal/services/containers/tests/resource_arm_kubernetes_cluster_network_test.go:
```diff
@@ -485,6 +486,60 @@ func testAccAzureRMKubernetesCluster_prefixedLoadBalancerProfile(t *testing.T) {
+					resource.TestCheckResourceAttr(data.ResourceName, "network_profile.0.load_balancer_profile.0.outbound_ip_address_ids.#", "1"),
+					resource.TestCheckResourceAttr(data.ResourceName, "network_profile.0.load_balancer_profile.0.effective_outbound_ips.#", "1"),
+				),
+			}
```
the compilation is failing here because of a missing `,`:


can we also add an import test between these two steps too, to check all the fields in the config are present in the state?

```suggestion
			},
			data.ImportStep(),
```

tombuildsstuff commented on azurerm/internal/services/containers/tests/resource_arm_kubernetes_cluster_network_test.go:
```diff
@@ -485,6 +486,60 @@ func testAccAzureRMKubernetesCluster_prefixedLoadBalancerProfile(t *testing.T) {
+func testAccAzureRMKubernetesCluster_changingLoadBalancerProfile(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_kubernetes_cluster", "test")
+	clientId := os.Getenv("ARM_CLIENT_ID")
+	clientSecret := os.Getenv("ARM_CLIENT_SECRET")
```
this is no longer needed, so we can remove this to fix the compilation error:

```suggestion
```

tombuildsstuff commented on azurerm/internal/services/containers/tests/resource_arm_kubernetes_cluster_network_test.go:
```diff
@@ -485,6 +486,60 @@ func testAccAzureRMKubernetesCluster_prefixedLoadBalancerProfile(t *testing.T) {
+
+func testAccAzureRMKubernetesCluster_changingLoadBalancerProfile(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_kubernetes_cluster", "test")
+	clientId := os.Getenv("ARM_CLIENT_ID")
```
this is no longer needed, so we can remove this to fix the compilation error:

```suggestion
```

aristosvo commented:
Acceptance test passed!
```
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
TF_ACC=1 go test -v ./azurerm/internal/services/containers/tests/ -run=TestAccAzureRMKubernetesCluster_changingLoadBalancerProfile -timeout 60m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMKubernetesCluster_changingLoadBalancerProfile
=== PAUSE TestAccAzureRMKubernetesCluster_changingLoadBalancerProfile
=== CONT  TestAccAzureRMKubernetesCluster_changingLoadBalancerProfile
--- PASS: TestAccAzureRMKubernetesCluster_changingLoadBalancerProfile (2014.92s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers/tests    2014.990s
```

tombuildsstuff commented:
Ignoring some expected failures due to capacity issues in Azure - the tests look good:

<img width="575" alt="Screenshot 2020-04-08 at 16 57 04" src="https://user-images.githubusercontent.com/666005/78799070-fab07380-79b9-11ea-9c9f-c9d3e5e3358b.png">

aristosvo commented:
Are these expected failures tracked somewhere? I’m only running targeted acceptance tests, but I was unaware of the failing ones.

tombuildsstuff commented:
@aristosvo unfortunately that varies test-run to test run due to capacity issues on the Azure side, so that's not easily possible - but we're checking these as we go
